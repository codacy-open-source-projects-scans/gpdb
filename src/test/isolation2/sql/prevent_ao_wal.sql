-- For AO/AOCO tables, their WAL records are only
-- generated for replication purposes (they are not used for crash
-- recovery because AO/AOCO table operations are crash-safe). To decrease
-- disk space usage and to improve performance of AO/AOCO operations, we
-- suppress generation of XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records when wal_level=minimal is
-- specified.
-- This test is supposed to ensure that XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records are not generated when
-- wal_level=minimal is set.
-- Because on mirrored cluster primary segments have replication slots
-- and that conflict with the wal_level=minimal GUC
-- we connect to coordinator in utility mode for AO/AOCO operations and
-- validate WAL records on the coordinator.

-- start_matchignore
-- m/pg_waldump: fatal: error in WAL record at */
-- m/.*The 'DISTRIBUTED BY' clause determines the distribution of data*/
-- m/.*Table doesn't have 'DISTRIBUTED BY' clause*/
-- end_matchignore

-- Create tables (AO, AOCO)
-1U: CREATE TABLE ao_foo (n int) WITH (appendonly=true);
-1U: CREATE TABLE aoco_foo (n int, m int) WITH (appendonly=true, orientation=column);

-- Switch WAL file
-1U: SELECT true FROM pg_switch_wal();
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
-1U: VACUUM ao_foo;
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
-1U: VACUUM aoco_foo;
-1Uq:

-- Validate wal records (mirrorless setting has alternative answer file for this since wal_level is already minimal)
! last_wal_file=$(psql -At -c "SELECT pg_walfile_name(pg_current_wal_lsn())" postgres) && pg_waldump ${last_wal_file} -p ${COORDINATOR_DATA_DIRECTORY}/pg_wal -r appendonly;

-- *********** Set wal_level=minimal **************
!\retcode gpconfig -c wal_level -v minimal --coordinatoronly;
-- Set max_wal_senders to 0 because a non-zero value requires wal_level >= 'archive'
!\retcode gpconfig -c max_wal_senders -v 0 --coordinatoronly;
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $COORDINATOR_DATA_DIRECTORY restart -w -t 600 -m fast;

-- Switch WAL file
-1U: SELECT true FROM pg_switch_wal();
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
-1U: VACUUM ao_foo;
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
-1U: VACUUM aoco_foo;

-- Validate wal records
! last_wal_file=$(psql -At -c "SELECT pg_walfile_name(pg_current_wal_lsn())" postgres) && pg_waldump ${last_wal_file} -p ${COORDINATOR_DATA_DIRECTORY}/pg_wal -r appendonly;

-1U: DROP TABLE ao_foo; 
-1U: DROP TABLE aoco_foo;

-- Reset wal_level
!\retcode gpconfig -r wal_level --coordinatoronly;
-- Reset max_wal_senders
!\retcode gpconfig -r max_wal_senders --coordinatoronly;
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $COORDINATOR_DATA_DIRECTORY restart -w -t 600 -m fast;
