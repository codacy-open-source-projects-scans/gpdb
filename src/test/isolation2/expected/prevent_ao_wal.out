-- For AO/AOCO tables, their WAL records are only
-- generated for replication purposes (they are not used for crash
-- recovery because AO/AOCO table operations are crash-safe). To decrease
-- disk space usage and to improve performance of AO/AOCO operations, we
-- suppress generation of XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records when wal_level=minimal is
-- specified.
-- This test is supposed to ensure that XLOG_APPENDONLY_INSERT and
-- XLOG_APPENDONLY_TRUNCATE WAL records are not generated when
-- wal_level=minimal is set.
-- Because on mirrored cluster primary segments have replication slots
-- and that conflict with the wal_level=minimal GUC
-- we connect to coordinator in utility mode for AO/AOCO operations and
-- validate WAL records on the coordinator.

GP_IGNORE: formatted by atmsort.pm
-- start_matchignore
-- m/pg_waldump: fatal: error in WAL record at */
-- m/.*The 'DISTRIBUTED BY' clause determines the distribution of data*/
-- m/.*Table doesn't have 'DISTRIBUTED BY' clause*/
-- end_matchignore
GP_IGNORE: defined new match expression

-- Create tables (AO, AOCO)
-1U: CREATE TABLE ao_foo (n int) WITH (appendonly=true);
CREATE TABLE
-1U: CREATE TABLE aoco_foo (n int, m int) WITH (appendonly=true, orientation=column);
CREATE TABLE

-- Switch WAL file
-1U: SELECT true FROM pg_switch_wal();
 bool 
------
 t    
(1 row)
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
INSERT 0 10
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
INSERT 0 10
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
DELETE 5
-1U: VACUUM ao_foo;
VACUUM
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
DELETE 5
-1U: VACUUM aoco_foo;
VACUUM
-1Uq: ... <quitting>

-- Validate wal records (mirrorless setting has alternative answer file for this since wal_level is already minimal)
! last_wal_file=$(psql -At -c "SELECT pg_walfile_name(pg_current_wal_lsn())" postgres) && pg_waldump ${last_wal_file} -p ${COORDINATOR_DATA_DIRECTORY}/pg_wal -r appendonly;
rmgr: Appendonly  len (rec/tot):    186/   186, tx:        554, lsn: 0/10000160, prev 0/10000118, desc: APPENDONLY_INSERT insert: rel 1663/16429/16384 seg/offset:0/0 len:136
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        555, lsn: 0/100003C8, prev 0/10000390, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:128/0 len:0
rmgr: Appendonly  len (rec/tot):    130/   130, tx:        555, lsn: 0/10000448, prev 0/10000400, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:0/0 len:80
rmgr: Appendonly  len (rec/tot):    130/   130, tx:        555, lsn: 0/100004D0, prev 0/10000448, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:128/0 len:80
rmgr: Appendonly  len (rec/tot):     50/    50, tx:          0, lsn: 0/100007C0, prev 0/10000790, desc: APPENDONLY_TRUNCATE truncate: rel 1663/16429/16384 seg/offset:0/136
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        557, lsn: 0/10000970, prev 0/10000938, desc: APPENDONLY_INSERT insert: rel 1663/16429/16384 seg/offset:1/0 len:0
rmgr: Appendonly  len (rec/tot):    138/   138, tx:        557, lsn: 0/10000A70, prev 0/10000A38, desc: APPENDONLY_INSERT insert: rel 1663/16429/16384 seg/offset:1/0 len:88
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        558, lsn: 0/10000BC0, prev 0/10000B90, desc: APPENDONLY_TRUNCATE truncate: rel 1663/16429/16384 seg/offset:0/0
rmgr: Appendonly  len (rec/tot):     50/    50, tx:          0, lsn: 0/100413F8, prev 0/100413C8, desc: APPENDONLY_TRUNCATE truncate: rel 1663/16429/16388 seg/offset:0/80
rmgr: Appendonly  len (rec/tot):     50/    50, tx:          0, lsn: 0/10041430, prev 0/100413F8, desc: APPENDONLY_TRUNCATE truncate: rel 1663/16429/16388 seg/offset:128/80
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        561, lsn: 0/10041600, prev 0/100415C8, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:1/0 len:0
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        561, lsn: 0/10041638, prev 0/10041600, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:129/0 len:0
rmgr: Appendonly  len (rec/tot):    114/   114, tx:        561, lsn: 0/10041738, prev 0/10041700, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:1/0 len:64
rmgr: Appendonly  len (rec/tot):    114/   114, tx:        561, lsn: 0/100417B0, prev 0/10041738, desc: APPENDONLY_INSERT insert: rel 1663/16429/16388 seg/offset:129/0 len:64
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        562, lsn: 0/10041908, prev 0/100418D8, desc: APPENDONLY_TRUNCATE truncate: rel 1663/16429/16388 seg/offset:0/0
rmgr: Appendonly  len (rec/tot):     50/    50, tx:        562, lsn: 0/10041940, prev 0/10041908, desc: APPENDONLY_TRUNCATE truncate: rel 1663/16429/16388 seg/offset:128/0
pg_waldump: fatal: error in WAL record at 0/10081FB8: invalid record length at 0/10081FE8: wanted 24, got 0


-- *********** Set wal_level=minimal **************
!\retcode gpconfig -c wal_level -v minimal --coordinatoronly;
(exited with code 0)
-- Set max_wal_senders to 0 because a non-zero value requires wal_level >= 'archive'
!\retcode gpconfig -c max_wal_senders -v 0 --coordinatoronly;
(exited with code 0)
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $COORDINATOR_DATA_DIRECTORY restart -w -t 600 -m fast;
(exited with code 0)

-- Switch WAL file
-1U: SELECT true FROM pg_switch_wal();
 bool 
------
 t    
(1 row)
-- Insert data (AO)
-1U: INSERT INTO ao_foo SELECT generate_series(1,10);
INSERT 0 10
-- Insert data (AOCO)
-1U: INSERT INTO aoco_foo SELECT generate_series(1,10), generate_series(1,10);
INSERT 0 10
-- Delete data and run vacuum (AO)
-1U: DELETE FROM ao_foo WHERE n > 5;
DELETE 5
-1U: VACUUM ao_foo;
VACUUM
-- Delete data and run vacuum (AOCO)
-1U: DELETE FROM aoco_foo WHERE n > 5;
DELETE 5
-1U: VACUUM aoco_foo;
VACUUM

-- Validate wal records
! last_wal_file=$(psql -At -c "SELECT pg_walfile_name(pg_current_wal_lsn())" postgres) && pg_waldump ${last_wal_file} -p ${COORDINATOR_DATA_DIRECTORY}/pg_wal -r appendonly;


-1U: DROP TABLE ao_foo;
DROP TABLE
-1U: DROP TABLE aoco_foo;
DROP TABLE

-- Reset wal_level
!\retcode gpconfig -r wal_level --coordinatoronly;
(exited with code 0)
-- Reset max_wal_senders
!\retcode gpconfig -r max_wal_senders --coordinatoronly;
(exited with code 0)
-- Restart QD
!\retcode pg_ctl -l /dev/null -D $COORDINATOR_DATA_DIRECTORY restart -w -t 600 -m fast;
(exited with code 0)
